src/vm-vcpu/src/vm.rs

```diff
@@ -0,0 +1,285 @@
+use crate::vcpu::mptable;
+use crate::vcpu::{self, KvmVcpu, VcpuState};
+use kvm_bindings::{kvm_pit_config, kvm_userspace_memory_region, KVM_PIT_SPEAKER_DUMMY};
+use kvm_ioctls::{Kvm, VmFd};
+use std::io;
+use std::sync::Arc;
+use std::thread;
+use std::thread::JoinHandle;
+use vm_device::device_manager::IoManager;
+use vm_memory::{Address, GuestMemory, GuestMemoryRegion};
+use vmm_sys_util::eventfd::EventFd;
+
+/// Defines the state from which a `KvmVm` is initialized.
+pub struct VmState {
+    pub num_vcpus: u8,
+}
+
+/// A KVM specific implementation of a Virtual Machine.
+///
+/// Provides abstractions for working with a VM. Once a generic Vm trait will be available,
+/// this type will become on of the concrete implementations.
+pub struct KvmVm {
+    fd: VmFd,
+    state: VmState,
+    // Only one of `vcpus` or `vcpu_handles` can be active at a time.
+    // To create the `vcpu_handles` the `vcpu` vector is drained.
+    // A better abstraction should be used to represent this behavior.
+    vcpus: Vec<KvmVcpu>,
```

Comment:
I'm not sure about storing the vCPUs in the `KvmVm` object. I don't see how they're related. One abstracts the hypervisor (`Vm`, currently `KvmVm` as it doesn't abstract it all the way) and the other abstracts something else entirely, the guest's CPUs. Moreover, the hypervisor is something you want to pass around later on in the execution of your program - for instance, devices need it at boot time to get activated (currently, that translates into registering the irqfd with KVM). I don't like sending the vCPUs too just for the sake of having fewer members in the `vmm` struct.

Reply:
I guess one way to look at this is `KvmVm` being an standalone abstraction that performs global and vcpu setup (and later start) based on KVM, behind an interface that provides as much convenience as possible. To that end, I think having the vcpus here makes sense; otherwise, as this becomes a separate component, each consumer would have to explicitly store them somewhere as well. It def seems useful to have accessors for some of the members, and I guess we'll get a better idea as we starting using these things and build on top of them.

Reply:
I think at the conceptual level the `vcpus` belong to the `Vm` abstraction because the VM should represent a machine with cpus & memory. Since there was no hurdle to actually move the ownership of the Vcpus to the Vm, I just went ahead and did that. This allows us to have `vm.run` that creates threads for the vcpus, and runs vcpu run. It is also the entity responsible for the lifetime of vcpus, and can do things such as joining the threads, and in the future kicking the vcpus when needed.